service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/container/CloneControlledAzureStorageContainerResourceFlight.java

```diff
@@ -0,0 +1,125 @@
+        WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_RESOURCE_ID,
+        WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID);
+
+    var flightBeanBag = FlightBeanBag.getFromObject(applicationContext);
+    var sourceResource =
+        inputParameters.get(WorkspaceFlightMapKeys.ResourceKeys.RESOURCE, ControlledResource.class);
+    var userRequest =
+        inputParameters.get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    var destinationWorkspaceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+
+    boolean mergePolicies =
+        Optional.ofNullable(
+                inputParameters.get(WorkspaceFlightMapKeys.MERGE_POLICIES, Boolean.class))
+            .orElse(false);
+    var cloningInstructions =
+        Optional.ofNullable(
+                inputParameters.get(
+                    WorkspaceFlightMapKeys.ControlledResourceKeys.CLONING_INSTRUCTIONS,
+                    CloningInstructions.class))
+            .orElse(sourceResource.getCloningInstructions());
+
+    final ControlledAzureStorageContainerResource sourceContainer =
+        sourceResource.castByEnum(WsmResourceType.CONTROLLED_AZURE_STORAGE_CONTAINER);
+    if (CloningInstructions.COPY_NOTHING == cloningInstructions) {
+      addStep(new SetNoOpContainerCloneResponseStep(sourceContainer));
+      return;
+    }
+
+    // Get the cloud context and store it in the working map
+    addStep(
+        new GetCloudContextStep(
+            destinationWorkspaceId,
+            CloudPlatform.AZURE,
+            flightBeanBag.getGcpCloudContextService(),
+            flightBeanBag.getAzureCloudContextService()),
+        RetryRules.shortDatabase());
+
+    // Flight plan
+    // 1. Check user has read access to source container
+    // 2. Gather controlled resource metadata for source object
+    // 3. Copy container definition to new container resource
+    // 3. If referenced, bail out as unsupported.
+    // 4. Copy files from source container to destination container (TODO)
+    addStep(
+        new CheckControlledResourceAuthStep(
+            sourceResource, flightBeanBag.getControlledResourceMetadataManager(), userRequest),
+        RetryRules.shortExponential());
+
+    if (mergePolicies) {
+      addStep(
+          new ClonePolicyAttributesStep(
+              sourceResource.getWorkspaceId(),
+              destinationWorkspaceId,
+              userRequest,
+              flightBeanBag.getTpsApiDispatch()));
+    }
+
+    addStep(
+        new RetrieveControlledResourceMetadataStep(
+            flightBeanBag.getResourceDao(),
+            sourceResource.getWorkspaceId(),
+            sourceResource.getResourceId()));
+
+    // TODO WOR-590 add step to copy source container metadata + attributes
+
+    if (CloningInstructions.COPY_REFERENCE == cloningInstructions) {
```

Comment:
Nit: it seems like we should validate this much earlier since there is no need to do anything else in this case. For instance, at input argument validation time. And it would be good to state in the Swagger documentation that COPY_REFERENCE is not supported.

Reply:
Also, there should be some test coverage of this and the COPY_NOTHING path.

Reply:
Fair. I'll leave this check here and add one at the controller level (belts and suspenders, etc.)